go/merkletree2/proof_verifier.go

```diff
@@ -100,3 +108,236 @@ func (m *MerkleProofVerifier) VerifyInclusionProof(ctx logger.ContextInterface,
 	// Success!
 	return nil
 }
+
+func (m *MerkleProofVerifier) computeSkipsHashForSeqno(s Seqno, skipsMap map[Seqno]Hash) (Hash, error) {
+	skipSeqnos := SkipPointersForSeqno(s)
+	skips := make([]Hash, len(skipSeqnos))
+	for i, s := range skipSeqnos {
+		skip, found := skipsMap[s]
+		if !found {
+			return nil, fmt.Errorf("the skipsMap in the proof does not contain necessary hash of seqno %v ", s)
+		}
+		skips[i] = skip
+	}
+	_, hash, err := m.cfg.Encoder.EncodeAndHashGeneric(skips)
+	if err != nil {
+		return nil, fmt.Errorf("Error encoding %+v: %v", skips, err)
+	}
+
+	return hash, nil
+}
+
+// computeFinalSkipPointersHashFromPath recomputes the SkipPointersHash for all
+// the seqnos on the SkipPointersPath(initialSeqno, finalSeqno), using the
+// information contained in the proof and returns the last one, as well as a
+// boolean indicating wether the proof is a full proof or is compressed (i.e. it
+// is part of a MerkleInclusionExtensionProof).
+func (m *MerkleProofVerifier) computeFinalSkipPointersHashFromPath(ctx logger.ContextInterface, proof *MerkleExtensionProof, initialSeqno Seqno, initialRootHash Hash, finalSeqno Seqno) (h Hash, isPartOfIncExtProof bool, err error) {
+	// This function is annotated with an inline example.
+
+	// We denote with Hi the hash of the RootMetadata Ri with seqno i.
+	// Example inputs:
+	// initialSeqno = 11
+	// initialRootHash = H11
+	// finalSeqno = 30
+	// the proof contains:
+	// - RootHashes = [ H8, H10, H14, H15, H24, H28, H29]
+	// - PreviousRootsNoSkips = [R12, R16, (R30)] Note that R30 here is optional.
+	//      We set isPartOfIncExtProof = true if it isn't there, but the output h is the same.
+
+	// Note that:
+	// SkipPointersPath(11,30) = [12,16,30]
+	// SkipPointersForSeqno(12) = [8,10,11]
+	// SkipPointersForSeqno(16) = [8,12,14,15]
+	// SkipPointersForSeqno(30) = [16,24,28,29]
+
+	rootHashMap := make(map[Seqno]Hash)
+	rootHashSeqnos, err := ComputeRootHashSeqnosNeededInExtensionProof(initialSeqno, finalSeqno)
+	// rootHashSeqnos = [8, 10, 14, 15, 24, 28, 29]
```

Comment:
I see, can you say this is the union of the 3 sets right above (if I understand?)

Reply:
No, it is a subset of the union. 
Note that 11 and 16 are not there because H11 = initialRootHash is an input to the proofVerifier and H16 can be computed from R16 and Hi for i in SkipPointersForSeqno(16) (i.e. H8,H12,H14,H15).

Reply:
Sorry, that is what i meant.